### PR TITLE
feat: Add pipeline action in rule, on push fire rule in release callback.

### PR DESCRIPTION
**Pull request by shuofan**
#### What this PR does / why we need it:
Add pipeline action in rule, on push fire rule in release callback.

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [1](https://erda.cloud/erda/dop/projects/387/issues/gantt?id=331900&iterationID=1354&pId=331880&type=TASK)
- [2](https://erda.cloud/erda/dop/projects/387/issues/gantt?id=331901&iterationID=1354&pId=331880&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  Add pipeline action in rule, on push fire rule in release callback.            |
| 🇨🇳 中文    |  规则触发支持 "git push" 事件；工作流支持 "执行流水线"      |

### CODE REVIEW

File: internal/apps/dop/endpoints/release.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ import (
 	"github.com/erda-project/erda-proto-go/dop/projectpipeline/pb"
 	"github.com/erda-project/erda/apistructs"
 	"github.com/erda-project/erda/internal/apps/dop/providers/projectpipeline"
+	pipelineaction "github.com/erda-project/erda/internal/apps/dop/providers/rule/actions/pipeline"
```
Comment (sfwn) [inline internal/apps/dop/endpoints/release.go:33]: `pipelineaction` has another meaning, represent for `pipeline action(job)`.
use another suitable name.

File: internal/apps/dop/endpoints/release.go [hunk 2]
```diff
@@ -102,6 +103,29 @@ func (e *Endpoints) ReleaseCallback(ctx context.Context, r *http.Request, vars m
 			continue
 		}
 
+		path, fileName := getSourcePathAndName(each)
+		go func() {
+			env := map[string]interface{}{
+				"content": req.Content,
```
Comment (sfwn) [inline internal/apps/dop/endpoints/release.go:109]: use typed variables, not literal.

File: internal/apps/dop/providers/rule/actions/pipeline/provider.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	reqPipeline := &apistructs.PipelineCreateRequest{
+		AppID:              req.AppID,
+		Branch:             req.RefName,
+		Source:             apistructs.PipelineSourceDice,
+		PipelineYmlSource:  apistructs.PipelineYmlSourceGittar,
+		PipelineYmlContent: req.PipelineYmlStr,
+		AutoRun:            true,
+		UserID:             req.UserID,
+	}
+	v2, err := p.pipeline.ConvertPipelineToV2(reqPipeline)
```
Comment (sfwn) [inline internal/apps/dop/providers/rule/actions/pipeline/provider.go:120]: I think these create logic is very similar to app pipeline. maybe code should be reused.

File: internal/apps/dop/providers/rule/actions/pipeline/provider.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	if err != nil {
+		logrus.Errorf("create pipeline failed, pipeline: %s, (%+v)", req.PipelineYmlStr, err)
+		return "", err
+	}
+
+	// store created pipeline id
+	return strconv.FormatUint(resp.ID, 10), nil
+}
+
+func (p *provider) GetPipelineDefinitionID(ctx context.Context, app *apistructs.ApplicationDTO, branch, path, name, strPipelineYml string) (definitionID string, err error) {
```
Comment (sfwn) [inline internal/apps/dop/providers/rule/actions/pipeline/provider.go:145]: any util function before?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/endpoints/release.go (hunk 2 above)
Comment (shuofan) [reply internal/apps/dop/endpoints/release.go:109]: done

File: internal/apps/dop/endpoints/release.go (hunk 1 above)
Comment (shuofan) [reply internal/apps/dop/endpoints/release.go:33]: done

File: internal/apps/dop/providers/rule/actions/pipeline/provider.go (hunk 3 above)
Comment (shuofan) [reply internal/apps/dop/providers/rule/actions/pipeline/provider.go:120]: they are in endpoints, any idea how we can avoid circular dependence.

File: internal/apps/dop/providers/rule/actions/pipeline/provider.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	PipelineYmlStr string
+	AppID          uint64
+	RefName        string
+	UserID         string
+	Path           string
+	FileName       string
+}
+
+func (p *provider) CreatePipeline(env map[string]interface{}) (string, error) {
+	gitPush, ok := env["git_push"]
```
Comment (sfwn) [inline internal/apps/dop/providers/rule/actions/pipeline/provider.go:80]: you should use `PipelineRuleEnv` here also.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/providers/rule/actions/pipeline/provider.go (hunk 5 above)
Comment (shuofan) [reply internal/apps/dop/providers/rule/actions/pipeline/provider.go:85]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**